### PR TITLE
Setup GitHub Actions to produce Windows binary on release tag creation

### PR DESCRIPTION
**Pull request by TheTechsTech**
- In reference to https://github.com/libuv/libuv/issues/1042, seems there should be one Windows location for binaries based on original and any changes.

### CODE REVIEW

File: .github/workflows/release-windows.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      - name: Upload the windows binary artifacts
+        uses: svenstaro/upload-release-action@v2
+        with:
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          file: uv-${{ github.event.release.tag_name }}.zip
+          asset_name: uv-${{ github.event.release.tag_name }}
+          tag: uv-${{ github.event.release.tag_name }}
+          overwrite: true
+          file_glob: true
+          body: "Window's binary - 'libuv' .dll, .lib, and .h headers."
```
Comment (bl-ue) [inline .github/workflows/release-windows.yml:41]:
```suggestion
          body: "Windows binary - 'libuv' .dll, .lib, and .h headers."
```

File: .github/workflows/release-windows.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    tags:
+      - '*'
+
+jobs:
+  windows:
+    name: windows-build
+    runs-on: windows-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: cmake build
```
Comment (bl-ue) [inline .github/workflows/release-windows.yml:14]:
```suggestion
      - name: CMake build
```

File: .github/workflows/release-windows.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        run: |
+             cmake .
+      - name: Setup MSBuild.exe
+        uses: microsoft/setup-msbuild@v1.0.2
+      - name: msbuild & test
```
Comment (bl-ue) [inline .github/workflows/release-windows.yml:19]:
```suggestion
      - name: MSBuild and tests
```

**Review by bl-ue — CHANGES_REQUESTED** [review]

**Review by justinmchase — APPROVED** [review]

**Comment by bnoordhuis** [comment]
@libuv/windows This needs your input. The changes look okay to my untrained eye.

**Comment by vtjnash** [comment]
I feel neutral-to-slightly-negatively towards maintaining binaries, but for no particular reason, just that I don't want to deal with them.

File: .github/workflows/release-windows.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  release:
+    tags:
+      - '*'
+
+jobs:
+  windows:
+    name: windows-build
+    runs-on: windows-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (vtjnash) [inline .github/workflows/release-windows.yml:13]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/release-windows.yml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        uses: microsoft/setup-msbuild@v1.0.2
+      - name: MSBuild and tests
+        run: |
+             msbuild INSTALL.vcxproj /p:Configuration=Release
+      - name: Create binary archive from build
+        run: |
+             mkdir libuv-${{ github.event.release.tag_name }}\bin
+             mkdir libuv-${{ github.event.release.tag_name }}\lib
+             move include libuv-${{ github.event.release.tag_name }}\
+             move Release/uv_a.lib libuv-${{ github.event.release.tag_name }}\lib\
```
Comment (vtjnash) [inline .github/workflows/release-windows.yml:27]: this is renamed now. `cmake install` might be more reliable for this?

File: .github/workflows/release-windows.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+             move Release/uv.exp libuv-${{ github.event.release.tag_name }}\lib\
+             move Release/uv.dll libuv-${{ github.event.release.tag_name }}\bin\
+             7z a libuv-${{ github.event.release.tag_name }}.zip libuv-${{ github.event.release.tag_name }}\* -tzip -y
+      - name: Upload the windows binary artifacts
+        uses: svenstaro/upload-release-action@v2
+        with:
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          file: libuv-${{ github.event.release.tag_name }}.zip
+          asset_name: libuv-${{ github.event.release.tag_name }}-windows
+          tag: libuv-${{ github.event.release.tag_name }}-windows
```
[the remark this answers is not in the document]
Comment (TheTechsTech) [reply .github/workflows/release-windows.yml:38]: This does not indicate windows only. 

I tried that with fork and the tag that shows is all wrong, not usable the way i way pulling it in.

File: .github/workflows/release-windows.yml (hunk 5 above)
Comment (vtjnash) [reply .github/workflows/release-windows.yml:27]: I thought `cmake` usually can be used as a driver for the underlying build program (and vice versa). I forget the exact invocation, but we use it already in several of the CI scripts. (indeed, we even now have a short-term upload of windows artifacts installed by cmake, as seen here https://github.com/libuv/libuv/actions/runs/3961926739)

But the lack of ongoing need is basically why we don't have this. It has been requested and proposed similar to this several times, but there has not seemed to be an active an on-going involvement by someone who can join the @libuv/windows team to help me maintain it.

Comment (TheTechsTech) [reply .github/workflows/release-windows.yml:27]: > as seen here https://github.com/libuv/libuv/actions/runs/3961926739

I have no formulary with cmake neither, I just used what the repo readme instructed. But it seems that instruction now is for installing on system, not for producing a zip file ready for next stage for uploading. Maybe for making **artifacts**. 

And that's the problem of making **artifacts** there is a stated expiration time to deal with, you need a team for that. 

Making a **release** has a different GitHub Action process, and separate location that's more part of repo itself. You would have to manually delete it. And you can easily programmability access the file, which should be zipped. 

I can help out on maintaining a Windows binary availability. But I see a different process, whoever decides a release is ready, just create a tag, doing so produces a **zipped** binary available for Windows, shows up under the **release page** along with the other **two** archive files.

I find that all Linux distributions include some version of Libuv. The project I'm working on, I build the current `Libuv Linux binary` and added directly to repo. Found out when using with **PHP FFI** it needed to be in users home directory or regular platform's OS directory. Windows had no restrictions. 

I decide to just use whatever Linux system has, deleted the files and GitHub Action to produce them, https://github.com/symplely/uv-ffi/commit/f9ca562b5174dca4b089fa1411fb8d2f33a730e8

File: .github/workflows/release-windows.yml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+on:
+  release:
+    types: [published]
+
+jobs:
+  windows:
+    name: windows-build
+    runs-on: windows-latest
+    steps:
+      - uses: actions/checkout@v3
```
Comment (saghul) [inline .github/workflows/release-windows.yml:12]: Use v4

File: .github/workflows/release-windows.yml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      - name: CMake build
+        run: |
+             cmake .
```
Comment (saghul) [inline .github/workflows/release-windows.yml:15]: Shouldn't you inlude the config with `--config Release` ?

**Comment by TheTechsTech** [comment]
I'm more in favor of https://github.com/libuv/libuv/pull/4489

**Pull request closed without merging**